src/Compilers/CSharp/Portable/Lowering/LambdaRewriter/LambdaRewriter.Analysis.cs

```diff
@@ -284,51 +222,67 @@ void RecordClosureScope(Scope innermost, Scope outermost, Closure closure)
                             }
                         }
                     }
-
                 }
             }
 
             /// <summary>
-            /// Walk up the scope tree looking for a closure.
+            /// Walk up the scope tree looking for a variable declaration.
             /// </summary>
-            /// <returns>
-            /// A tuple of the found <see cref="Closure"/> and the <see cref="Scope"/> it was found in.
-            /// </returns>
-            public static (Closure, Scope) GetVisibleClosure(Scope startingScope, MethodSymbol closureSymbol)
+            public static Scope GetVariableDeclarationScope(Scope startingScope, Symbol variable)
             {
+                if (variable is ParameterSymbol p && p.IsThis)
+                {
+                    return null;
+                }
+
                 var currentScope = startingScope;
                 while (currentScope != null)
                 {
-                    foreach (var closure in currentScope.Closures)
+                    if (variable.Kind == SymbolKind.Parameter || variable.Kind == SymbolKind.Local)
                     {
-                        if (closure.OriginalMethodSymbol == closureSymbol)
+                        if (currentScope.DeclaredVariables.Contains(variable))
                         {
-                            return (closure, currentScope);
+                            return currentScope;
+                        }
+                    }
+                    else
+                    {
+                        if (currentScope.Closures.Contains(c => variable == c.OriginalMethodSymbol))
```

Comment:
Are there concerns about this being a hot method, with respect to the lambda allocation here? (a quick glance says that this method is called over every captured var in every closure)

(I'm not sure how strict we are about LINQ/perf, and the rules surrounding when it's okay to use and not)